src/module.ts

```diff
@@ -2484,115 +2484,19 @@ export class Module {
-          }
-          case <u32>TypeRef.F64: {
-            return this.f64(binaryen._BinaryenConstGetValueF64(expr));
-          }
-          case <u32>TypeRef.V128: {
-            // TODO
-            return 0;
-          }
-          // Not possible to clone an externref as it is opaque
-          case <u32>TypeRef.Externref: {
-            return 0;
-          }
-          default: {
-            throw new Error("concrete type expected");
-          }
-        }
-      }
-      case ExpressionId.LocalGet: {
-        return binaryen._BinaryenLocalGet(this.ref,
-          binaryen._BinaryenLocalGetGetIndex(expr),
-          binaryen._BinaryenExpressionGetType(expr)
-        );
-      }
-      case ExpressionId.GlobalGet: {
-        let globalName = binaryen._BinaryenGlobalGetGetName(expr);
-        if (!globalName) break;
-        return binaryen._BinaryenGlobalGet(
-          this.ref, globalName, binaryen._BinaryenExpressionGetType(expr)
-        );
-      }
-      case ExpressionId.Load: {
-        if (!(nested1 = this.cloneExpression(
-          binaryen._BinaryenLoadGetPtr(expr), noSideEffects, maxDepth)
-        )) {
-          break;
-        }
-        return (
-          binaryen._BinaryenLoadIsAtomic(expr)
-            ? binaryen._BinaryenAtomicLoad(this.ref,
-                binaryen._BinaryenLoadGetBytes(expr),
-                binaryen._BinaryenLoadGetOffset(expr),
-                binaryen._BinaryenExpressionGetType(expr),
-                nested1
-              )
-            : binaryen._BinaryenLoad(this.ref,
-                binaryen._BinaryenLoadGetBytes(expr),
-                binaryen._BinaryenLoadIsSigned(expr),
-                binaryen._BinaryenLoadGetOffset(expr),
-                binaryen._BinaryenLoadGetAlign(expr),
-                binaryen._BinaryenExpressionGetType(expr),
-                nested1
-              )
-        );
-      }
-      case ExpressionId.Unary: {
-        if (!(nested1 = this.cloneExpression(
-          binaryen._BinaryenUnaryGetValue(expr), noSideEffects, maxDepth)
-        )) {
-          break;
-        }
-        return binaryen._BinaryenUnary(
-          this.ref, binaryen._BinaryenUnaryGetOp(expr), nested1
-        );
-      }
-      case ExpressionId.Binary: {
-        if (!(nested1 = this.cloneExpression(
-          binaryen._BinaryenBinaryGetLeft(expr), noSideEffects, maxDepth)
-        )) {
-          break;
-        }
-        if (!(nested2 = this.cloneExpression(
-          binaryen._BinaryenBinaryGetRight(expr), noSideEffects, maxDepth)
-        )) {
-          break;
-        }
-        return binaryen._BinaryenBinary(
-          this.ref, binaryen._BinaryenBinaryGetOp(expr), nested1, nested2
-        );
-      }
+      case ExpressionId.LocalGet:
+      case ExpressionId.GlobalGet:
+      case ExpressionId.Const:
+      case ExpressionId.Unreachable:
+      case ExpressionId.RefNull: return this.copyExpression(expr);
```

Comment:
`ExpressionId.Load`?

Reply:
Expressions here only have constant immediates (if any) but don't contain subexpressions. Loads on the other hand take pointer expressions that can again be arbitrarily complex. Matches what `cloneExpression` did with `maxDepth=0`.

Reply:
I see. Makes sense. Could you add comment about this?

Reply:
Btw how about `ExpressionId.Nop`, `ExpressionId.DataDrop`, `ExpressionId.MemorySize`, `ExpressionId.RttCanon`, `ExpressionId.StructGet`, `ExpressionId.ArrayLen` ?

Reply:
`struct.get` and `array.len` take an expression operand, but the others should be good :)

Reply:
you're right